include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -17,55 +17,110 @@ namespace hlsl
+    static OrientedEtas<T> create(scalar_type NdotI, T eta)
     {
-        const bool backside = NdotI < 0.0;
-        const float rcpEta = 1.0 / eta;
-        orientedEta = backside ? rcpEta : eta;
-        rcpOrientedEta = backside ? eta : rcpEta;
-        return backside;
+        OrientedEtas<T> retval;
+        retval.backside = NdotI < hlsl::promote<T>(0.0);
+        const T rcpEta = hlsl::promote<T>(1.0) / eta;
+        retval.value = retval.backside ? rcpEta : eta;
+        retval.rcp = retval.backside ? eta : rcpEta;
+        return retval;
     }
+
+    static T diffuseFresnelCorrectionFactor(T n, T n2)
+    {
+        // assert(n*n==n2);
+        vector<bool,vector_traits<T>::Dimension> TIR = n < (T)1.0;
+        T invdenum = nbl::hlsl::mix<T>(hlsl::promote<T>(1.0), hlsl::promote<T>(1.0) / (n2 * n2 * (hlsl::promote<T>(554.33) - 380.7 * n)), TIR);
+        T num = n * nbl::hlsl::mix<T>(hlsl::promote<T>(0.1921156102251088), n * 298.25 - 261.38 * n2 + 138.43, TIR);
+        num += nbl::hlsl::mix<T>(hlsl::promote<T>(0.8078843897748912), hlsl::promote<T>(-1.67), TIR);
+        return num * invdenum;
+    }
+
+    T value;
+    T rcp;
+    bool backside;
 };
 
-template<>
-struct orientedEtas<float32_t3>
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T> || is_vector_v<T>)
+struct OrientedEtaRcps
 {
-    static bool __call(NBL_REF_ARG(float32_t3) orientedEta, NBL_REF_ARG(float32_t3) rcpOrientedEta, float NdotI, float32_t3 eta)
+    using scalar_type = typename vector_traits<T>::scalar_type;
+
+    static OrientedEtaRcps<T> create(scalar_type NdotI, T eta)
     {
-        const bool backside = NdotI < 0.0;
-        const float32_t3 rcpEta = (float32_t3)1.0 / eta;
-        orientedEta = backside ? rcpEta:eta;
-        rcpOrientedEta = backside ? eta:rcpEta;
-        return backside;
+        OrientedEtaRcps<T> retval;
+        retval.backside = NdotI < hlsl::promote<T>(0.0);
+        const T rcpEta = hlsl::promote<T>(1.0) / eta;
+        retval.value = retval.backside ? eta : rcpEta;
+        retval.value2 = retval.value * retval.value;
+        return retval;
     }
+
+    static T diffuseFresnelCorrectionFactor(T n, T n2)
+    {
+        return OrientedEtas<T>::diffuseFresnelCorrectionFactor(n, n2);
+    }
+
+    T value;
+    T value2;
+    bool backside;
 };
+
 }
 
-template<typename T NBL_FUNC_REQUIRES(is_scalar_v<T> || is_vector_v<T>)
-bool getOrientedEtas(NBL_REF_ARG(T) orientedEta, NBL_REF_ARG(T) rcpOrientedEta, scalar_type_t<T> NdotI, T eta)
+template<typename T NBL_PRIMARY_REQUIRES(concepts::Vectorial<T> && vector_traits<T>::Dimension == 3)
+struct Reflect
 {
-    return impl::orientedEtas<T>::__call(orientedEta, rcpOrientedEta, NdotI, eta);
-}
+    using this_t = Reflect<T>;
+    using vector_type = T;
+    using scalar_type = typename vector_traits<T>::scalar_type;
 
+    static this_t computeNdotI(NBL_CONST_REF_ARG(vector_type) I, NBL_CONST_REF_ARG(vector_type) N)
+    {
+        return hlsl::dot<vector_type>(N, I);
+    }
```

Comment:
ok maybe the `compute` should be called `recompute`, the should all be void.

The idea is that I can fill (be it by `create` or manually) the struct from pure independent variables, then call `recompute` to get all the optionals if I don't already have them

Even if `create(...IndependentVars...)`  calls `recomputeXYZ` I can get those redundant functions optimized out by simply overwriting their result
```cpp
auto refl = Reflect<float>::create(Incident,Normal);
refl.NdotI = cosView;
```

this is a general pattern I want to use, to avoid having 123555939 overloads of a constructor for every occasion